### ISSUE TITLE
Clarifications of ASYNC119 scope and details

### ISSUE THREAD

**Issue by RomanValov** [Zac-HD/flake8-trio#395]
The issue is considered to be a follow up of discussion at #238.

0. The rules page states ( https://flake8-async.readthedocs.io/en/stable/rules.html#async119 ):

   > `yield` in context manager in async generator is unsafe, the cleanup may be delayed until `await` is no longer allowed

   But the rule is triggered for sync context managers too. But sync CMs can't have any `await`s inside. So the description is confusing.

   FTR, error message displayed by `flake8` is:

   > Yield in contextmanager in async generator might not trigger cleanup. Use `@asynccontextmanager` or refactor.

1. As per comments in original pull request, there are basically two distinct issues / leves of concern mixed into the issue:

   a. Generator cleanup may be performed by garbage collector and may not be performed promptly.
   b. Generator cleanup may be performed in context where coroutine runner is not available and `await`s are doomed to be skipped/raise errors.

   IMO these two cases should be split into distinct error because case a is less severe than case b. It may be desired to "ignore" them independently. Moreover as demonstrated by snippet in [my comment](https://github.com/python-trio/flake8-async/pull/238#issuecomment-3104204264), case a applies to purely sync code with sync generators/sync context managers. So maybe such rule should be pushed towards more generic linters, i.e. `flake8` on its own.

2. Back to statement that generator cleanup may run in context where coroutine runner is not available. As per my understanding it should not happen with asyncio due to shutdown asyncgens machinery. Well, at least if `asyncio.run` is used. It would be nice to have a demonstrating snippet and clarify conditions under which it could happen (i.e. use trio, use pypy, etc.). So we have reproducible example everyone could check and developers may decide if they should bother for their case. Snippets posted by Zac and me on original pull request discussion both demonstrate case a, but not case b. In case issue only applies to non-asyncio libraries, the rule could be adjusted like it was done with ASYNC102.

3. Last but not least. In my code base I have snippets like:

   ```python
   from contextlib import suppress

   async def iterate():
     with suppress(SomeError):
       yield
       yield
   ```

   I think the snippet is perfectly safe (even omitting my other concerns), because implementation of `suppress.__exit__` only inspects exception in case it was thrown and do nothing otherwise. However it seems there is no way to instruct flake8-async about it other then adding "noqa" directives. There is `--transform-async-generator-decorators` but it's too broad for my case because it marks overall generator as safe, but not exact context manager used inside generator.

**Comment by RomanValov** [comment]
CC @alicederyn

**Comment by alicederyn** [comment]
That await isn't associated with the context manager though, and it would be flagged up separately because we check for await in finally blocks. There's no way to "hide" an await within a synchronous context manager.

**Comment by jakkdl** [comment]
yeah sorry, I'll clarify a bit more:

ASYNC102 (await-in-finally-or-canceled) no longer triggers for asyncio since #393, since the edge cancellation semantics means that `await` is sometimes fine and necessary.
For trio, with its level cancellation semantics (i.e. once canceled every subsequent `await` will raise `Cancelled`) it's never fine to `await` in `finally`/`except Cancelled` unless you shield it from outside cancellations:
```python
try:
    ...
finally:
    with trio.CancelScope(shield=True):
        await my_cleanup_function()  # perfectly fine, does not trigger ASYNC102
```

So we cannot rely on warnings for `await` in `finally`/`except` for warning about this entirely separate problem where the async gen can be stuck inside the `try` and not enter the except/finally until much later - where the run loop theoretically may no longer be running.

But ASYNC119 would be somewhat misleading here, triggering on the CM and not on the `await` in the finally, so we should probably add some more logic to ASYNC119 for this case.
